### ISSUE TITLE
Feature Request: Get or Insert Future

### ISSUE THREAD

**Issue by Dessix** [moka-rs/moka#15]
The ability to get-or-insert atomically would be nice for preventing multiple requests for the same resource from being created to populate the cache. Otherwise, there's a race condition between get and insert where another task could attempt to get as well, and also see an empty record to populate.

**Comment by tatsuya6502** [comment]
Thanks for your suggestion. I think this would be a good feature to add. I will work on it once I finish working on another feature (#12).

Which Moka `Cache` implementation are you going to use? [moka::sync::Cache](https://docs.rs/moka/0.3.0/moka/sync/struct.Cache.html) or [moka::future::Cache](https://docs.rs/moka/0.3.0/moka/future/struct.Cache.html)?

**Comment by Dessix** [comment]
Futures- thanks ^^

**Comment by tatsuya6502** [comment]
Hi @Dessix 

I created [a topic branch](https://github.com/moka-rs/moka/tree/get_or_insert_with) and added the following methods to the future cache:

```rust
async fn get_or_insert_with(&self, key: K, init: impl Future<Output = V>) -> V
```

- This method takes a key and a future to initialize the value.
- If the key does not exist in the cache and no other `get_or_insert_with` or `get_or_try_insert_with` method is already running for the key, it awaits for the future, inserts the value returned from the future to the cache, and returns the value to the caller.
- If other method is already running (the future is being awaited), it waits for the other method to complete and returns the value.

```rust
async fn get_or_try_insert_with<F>(&self, key: K, init: F) -> Result<V, Arc<Box<dyn Error>>>
where
    F: Future<Output = Result<V, Box<dyn Error>>>
```

- This method is similar to `get_or_insert_with` but the future resolves into `Result<V, Box<dyn std::error::Error>>` instead of `V`.
- A value is inserted to the cache only when the future resolves into `Ok(V)`.

Please check the unit tests out as they will show you how to use them ([test1][test1] and [test2][test2]). Let me know what you think.

[test1]: https://github.com/moka-rs/moka/blob/ccbb699ffef52d69b612cbfc6356666e648f19de/src/future/cache.rs#L894
[test2]: https://github.com/moka-rs/moka/blob/ccbb699ffef52d69b612cbfc6356666e648f19de/src/future/cache.rs#L979

**Comment by tatsuya6502** [comment]
Published Moka [v0.5.0](https://crates.io/crates/moka/0.5.0) to crates.io.
